### PR TITLE
#99-kmp-redux.

### PR DESCRIPTION
**Pull request by bogdan-evtushenko**

### CODE REVIEW

File: app/src/main/java/com/bogdan/codeforceswatcher/features/MainActivity.kt [hunk 1]
```diff
@@ -44,70 +43,55 @@ class MainActivity : AppCompatActivity(), StoreSubscriber<UIState> {
         initViews()
     }
 
-    override fun onStart() {
-        super.onStart()
-        store.subscribe(this) { state ->
-            state.skipRepeats { oldState, newState -> oldState.ui == newState.ui }
-                    .select { it.ui }
+    private fun updateFragment() {
+        val bottomNavSelectedItemId = when (selectedHomeTab) {
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/MainActivity.kt]: Can be extracted to function.

File: app/src/main/java/com/bogdan/codeforceswatcher/features/MainActivity.kt [hunk 2]
```diff
@@ -44,70 +43,55 @@ class MainActivity : AppCompatActivity(), StoreSubscriber<UIState> {
...
+        tvPageTitle.text = getString(selectedHomeTab.titleId)
         toolbar.collapseActionView()
 
         if (bottomNavigation.selectedItemId != bottomNavSelectedItemId) {
             bottomNavigation.selectedItemId = bottomNavSelectedItemId
         }
 
-        when (state.selectedHomeTab) {
-            UIState.HomeTab.USERS -> {
+        when (selectedHomeTab) {
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/MainActivity.kt]: This one too and cases can be one-lined.

File: app/src/main/java/com/bogdan/codeforceswatcher/features/MainActivity.kt [hunk 3]
```diff
@@ -130,8 +114,7 @@ class MainActivity : AppCompatActivity(), StoreSubscriber<UIState> {
         searchViewItem?.isVisible = false
 
         fab.setOnClickListener {
-            val intent =
-                    Intent(Intent.ACTION_VIEW).setData(Uri.parse(CONTESTS_LINK))
+            val intent = Intent(Intent.ACTION_VIEW).setData(Uri.parse(CONTESTS_LINK))
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/MainActivity.kt]: 👍

File: app/src/main/java/com/bogdan/codeforceswatcher/handlers/NotificationMiddleware.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    override fun handle(notificationData: List<Pair<String, Int>>) {
+        val notificationText = buildNotificationText(notificationData)
+
+        if (notificationText.isNotEmpty()) {
+            showNotification(CwApp.app, notificationText)
+        }
+    }
+
+    private fun buildNotificationText(notificationData: List<Pair<String, Int>>) =
+            notificationData.joinToString(separator = "\n") { ratingChange ->
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/handlers/NotificationMiddleware.kt]: Strange formatting here.

File: common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/add_user/redux/requests/AddUserRequests.kt [hunk 5]
```diff
@@ -29,7 +27,7 @@ class AddUserRequests {
                 user.id = DatabaseQueries.Users.insert(user)
                 store.dispatch(Success(user))
             } else {
-                store.dispatch(Failure(CwApp.app.getString(R.string.user_already_added)))
+                store.dispatch(Failure("User already added"))
```
Comment (yevhenii-kanivets) [inline common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/add_user/redux/requests/AddUserRequests.kt]: Shouldn't it be a localized string?

**Comment by yevhenii-kanivets** [comment]
@bogdan-evtushenko I've quickly gone through all the files and refactored what looked strange to me. After this refactoring, I've bound common module to iOS project. Not Den should be able to use common module in iOS project. (Maybe some changes will be needed on top of what we have now).

Can you, please, take a look on my commits and test Android app. @MatyashDen it would be very helpful if you test the app quickly too 🙏

**Comment by MatyashDen** [comment]
Works great!

**Pull request merged**